File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/DailyTemperatures.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayDeque;

public class DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        ArrayDeque<Integer> stack = new ArrayDeque<>();
        int[] result = new int[temperatures.length];

        for (int day = 0; day < temperatures.length; day++) {
            while (!stack.isEmpty() && temperatures[stack.peekLast()] < temperatures[day]) {
                int prevDay = stack.removeLast();
                result[prevDay] = day - prevDay;
            }
            stack.addLast(day);
        }

        return result;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/dailyTemperatures.kt
package com.only4.algorithm.version4k.leetcode

fun dailyTemperatures(temperatures: IntArray): IntArray {
    val stack = ArrayDeque<Int>()
    val result = IntArray(temperatures.size)

    temperatures.forEachIndexed { day, temperature ->
        while (stack.isNotEmpty() && temperatures[stack.last()] < temperature) {
            val prevDay = stack.removeLast()
            result[prevDay] = day - prevDay
        }
        stack.addLast(day)
    }

    return result
}
